drivers/input/input_adc_keys.c

```diff
@@ -0,0 +1,228 @@
+	int32_t sample_mv, closest_mv;
+	uint32_t diff, closest_diff = UINT32_MAX;
+	const struct adc_keys_code_config *code_cfg;
+	struct adc_keys_key_state *key_state;
+	uint16_t key_code;
+
+	sample_mv = adc_keys_read(dev);
+
+	/*
+	 * Find the closest key press threshold to the sample value.
+	 */
+
+	for (uint8_t i = 0; i < cfg->code_cnt; i++) {
+		diff = abs(sample_mv - cfg->code_cfg[i].press_mv);
+		if (diff < closest_diff) {
+			closest_diff = diff;
+			closest_mv = cfg->code_cfg[i].press_mv;
+		}
+	}
+
+	diff = abs(sample_mv - cfg->keyup_mv);
+	if (diff < closest_diff) {
+		closest_diff = diff;
+		closest_mv = cfg->keyup_mv;
+	}
+
+	LOG_DBG("sample=%d mV, closest=%d mV, diff=%d mV", sample_mv, closest_mv, closest_diff);
+
+	/*
+	 * Update cached key states according to the closest key press threshold.
+	 *
+	 * Note that multiple keys may have the same press threshold, which is
+	 * the mixed voltage that these keys are simultaneously pressed.
+	 */
+
+	for (uint8_t i = 0; i < cfg->code_cnt; i++) {
+		code_cfg = &cfg->code_cfg[i];
+		key_state = &cfg->key_state[code_cfg->key_index];
+
+		/*
+		 * Only update curr_state if the key is pressed to prevent
+		 * being overwritten by another threshold configuration.
+		 */
+		if (closest_mv == code_cfg->press_mv) {
+			key_state->curr_state = true;
+		}
+	}
+
+	/*
+	 * Report the key event if the key state has changed.
+	 */
+
+	for (uint8_t i = 0; i < cfg->key_cnt; i++) {
+		key_state = &cfg->key_state[i];
+		key_code = cfg->key_code[i];
+
+		if (key_state->last_state != key_state->curr_state) {
+			LOG_DBG("Report event %s %d, code=%d", dev->name, key_state->curr_state,
+				key_code);
+			input_report_key(dev, key_code, key_state->curr_state, true, K_FOREVER);
+			key_state->last_state = key_state->curr_state;
+		}
+
+		/*
+		 * Reset the state so that it can be updated in the next
+		 * iteration.
+		 */
+		key_state->curr_state = false;
+	}
+}
+
+static void adc_keys_work_handler(struct k_work *work)
+{
+	struct k_work_delayable *dwork = CONTAINER_OF(work, struct k_work_delayable, work);
+	struct adc_keys_data *data = CONTAINER_OF(dwork, struct adc_keys_data, dwork);
+	const struct device *dev = data->self;
+	const struct adc_keys_config *cfg = dev->config;
+
+	adc_keys_process(dev);
+
+	k_work_schedule(&data->dwork, K_MSEC(cfg->sample_period_ms));
+}
+
+static int adc_keys_init(const struct device *dev)
+{
+	const struct adc_keys_config *cfg = dev->config;
+	struct adc_keys_data *data = dev->data;
+	int ret;
+
+	if (!device_is_ready(cfg->channel.dev)) {
+		LOG_ERR("ADC controller device %s not ready", cfg->channel.dev->name);
+		return -ENODEV;
+	}
+
+	ret = adc_channel_setup_dt(&cfg->channel);
+	if (ret) {
+		LOG_ERR("ADC channel setup failed %d", ret);
+		return ret;
+	}
+
+	adc_sequence_init_dt(&cfg->channel, &data->seq);
+
+	data->self = dev;
+	k_work_init_delayable(&data->dwork, adc_keys_work_handler);
+
+#if defined(CONFIG_INPUT_LOG_LEVEL_DBG)
```

Comment:
```suggestion
if (IS_ENABLED((CONFIG_INPUT_LOG_LEVEL_DBG))
```

Reply:
Why? It should be completely removed in compile time if not enabled

Reply:
Linker should drop this part anyway. You can double-check that by comparing the binaries.

Reply:
This seems a bit too dependent on the compiler's optimization level. Wouldn't using the preprocessor be a more recommended approach, as discussed in #53200?

Reply:
It depends. In that case, the compiler will catch the errors in that for loop even though `CONFIG_INPUT_LOG_LEVEL_DBG` is disabled. As it's an init function it does not have much impact on the code even though the linker would not drop it.

Reply:
Oh, that's enlightening. Thank you for your suggestion. Updated.

Reply:
thanks for catching this @bbilas -- there's a bit in the coding style guide: https://docs.zephyrproject.org/latest/kernel/util/index.html#c.IS_ENABLED, yes it does depend on compiler optimization but that's a fairly common theme in the code base and at this point just something understood and utilized